### PR TITLE
add support of gather_object

### LINKED ISSUE #235: stronger gather implement
`gather` is crucial in evaluating model. current implementation assumes each tensor across devices has same size, which is very unlikely to happen in the real case. calling `pad_across_processes` results to additional 0s in the output, leading to incorrect results. 

```python
def _gpu_gather_one(tensor, same_size = True):
    if tensor.ndim == 0:
        tensor = tensor.clone()[None]

    if not same_size:
        size = torch.tensor(tensor.shape[0])[None]
        size_list = [torch.zeros(1) for _ in range(torch.distributed.get_world_size())]
        torch.distributed.all_gather(size_list, size)

    output_tensors = [torch.zeros_like(tensor) for _ in range(torch.distributed.get_world_size())]
    torch.distributed.all_gather(output_tensors, tensor)

    if not same_ssize:
        output_tensors = [i[:j] for i, j in zip(output_tensors, size_list)]
    return torch.cat(output_tensors, dim=0)
```

one another workaround is to convert results into `list` and use `gather_object` to gather. However, `gather_object` is also not provided in accelerate.

### PR DESCRIPTION
**Pull request by ZhiyuanChen**
fixes #235

### CODE REVIEW

File: src/accelerate/utils.py [hunk 1]
```diff
@@ -345,6 +345,40 @@ def gather(tensor):
...
+
+_cpu_gather_object = _gpu_gather_object
+
+
+def gather_object(object: Any):
+    """
+    Recursively gather object in a nested list/tuple/dictionary of objects from all devices.
+
+    Args:
+        object (nested list/tuple/dictionary of :obj:`Any`):
```
Comment (sgugger) [inline src/accelerate/utils.py:366]:
```suggestion
        object (nested list/tuple/dictionary of any object):
```

File: src/accelerate/utils.py [hunk 2]
```diff
@@ -345,6 +345,40 @@ def gather(tensor):
         return tensor
 
 
+
+def _gpu_gather_object(object: Any):
+    def _gpu_gather_object_one(object: Any):
+        output_objects = [None for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather_object(output_objects, object)
+        return [i for j in output_objects for i in j]
```
Comment (sgugger) [inline src/accelerate/utils.py:353]: I don't understand this line. This presupposes that every element in `output_objects` is iterable, but `object` is supposed to be a primitive object at this point (tensor or other if we remove `error_on_type=True`).

**Review by sgugger — COMMENTED** [review]
Thanks a lot for working on this! Could you explain a bit more the inner work of your implementation?

**Comment by ZhiyuanChen** [comment]
> I don't understand this line. This presupposes that every element in output_objects is iterable, but object is supposed to be a primitive object at this point (tensor or other if we remove error_on_type=True).

Should we separate `gather_object` to two functions, `gather_iterable` perform `[i for j in output_objects for i in j]` so it works much like original `gather`, while `gather_object` doesn not perform concatenation so it has better support for every objects

**Comment by sgugger** [comment]
It can't really concatenate if the tensors are not of the same size, so `gather_object` should not concatenate IMO.

**Comment by ZhiyuanChen** [comment]
> It can't really concatenate if the tensors are not of the same size, so `gather_object` should not concatenate IMO.

Yes we can leave it to the user to concat manually.

**Comment by ZhiyuanChen** [comment]
I have resolved all conversations, could you please review it again?

**Review by sgugger — APPROVED** [review]
Thanks for addressing all comments, this looks good to merge!

**Comment by ZhiyuanChen** [comment]
> Thanks for addressing all comments, this looks good to merge! Could you just run `make style` on your branch to fix the quality failure?

Sorry for this late reply.... I believe I've resolved the quality issue.

File: src/accelerate/utils.py [hunk 3]
```diff
@@ -350,6 +350,39 @@ def gather(tensor):
         return tensor
 
 
+def _gpu_gather_object(object: Any):
+    def _gpu_gather_object_one(object: Any):
+        output_objects = [None for _ in range(state.num_processes)]
```
Comment (sgugger) [inline src/accelerate/utils.py:355]:
```suggestion
        output_objects = [None for _ in range(AcceleratorState().num_processes)]
```
`state` is not defined.

File: src/accelerate/utils.py [hunk 4]
```diff
@@ -350,6 +350,39 @@ def gather(tensor):
...
+        torch.distributed.all_gather_object(output_objects, object)
+        return output_objects
+
+    return recursively_apply(_gpu_gather_one, tensor)
```
Comment (sgugger) [inline src/accelerate/utils.py:359]:
```suggestion
    return recursively_apply(_gpu_gather_object_one, object)
```

**Review by sgugger — COMMENTED** [review]
Here is how to solve the last quality issues (variables not defined).

**Comment by ZhiyuanChen** [comment]
Hmmmm, is it possible for us to determine the type of input data and auto call corresponding gather function? I just noticed [torchvison](https://github.com/pytorch/vision/blob/main/references/detection/utils.py#L70-L83) uses `gather_object` at all time

**Comment by sgugger** [comment]
That's too magical for Accelerate's philosophy, users should pick which function they want to use.

All green now so we can merge your PR :-)

**Comment by ZhiyuanChen** [comment]
> That's too magical for Accelerate's philosophy, users should pick which function they want to use.

Haha, thats the reason why i love this repo so much. Users should be trained programmer and should know what they're doing/supposed to do. 

> All green now so we can merge your PR :-) Thanks again!

Thank you for all the review and comments~

**Pull request merged**